onnxruntime/contrib_ops/cpu/transformers/sampling.cc

```diff
@@ -0,0 +1,140 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// there's no way to use a raw pointer as the copy destination with std::copy_n
+// (which gsl::copy uses with span::data() which returns a raw pointer) with the 14.11 toolset
+// without generating a 4996 warning. going through an iterator is way too much overhead so turn off the warning.
+#ifdef _MSC_VER
+#pragma warning(push)
+#pragma warning(disable : 4996)
+#endif
+
+#include "core/framework/op_kernel_context_internal.h"
+#include "core/framework/utils.h"
+#include "contrib_ops/cpu/transformers/sampling.h"
+#include "contrib_ops/cpu/transformers/logits_processor.h"
+#include "contrib_ops/cpu/transformers/sequences.h"
+#include "contrib_ops/cpu/transformers/dump_tensor.h"
+#include "contrib_ops/cpu/transformers/greedy_search_impl_gpt.h"
+
+using namespace ONNX_NAMESPACE;
+using namespace onnxruntime::common;
+
+namespace onnxruntime {
+namespace contrib {
+
+#define REGISTER_KERNEL_TYPED(T)                                  \
+  ONNX_OPERATOR_TYPED_KERNEL_EX(                                  \
+      Sampling,                                                   \
+      kMSDomain,                                                  \
+      1,                                                          \
+      T,                                                          \
+      kCpuExecutionProvider,                                      \
+      (*KernelDefBuilder::Create())                               \
+          .TypeConstraint("T", DataTypeImpl::GetTensorType<T>()), \
+      transformers::Sampling);
+
+REGISTER_KERNEL_TYPED(float)
+
+namespace transformers {
+
+void Sampling::Init(const OpKernelInfo& info) {
+  parameters_.ParseFromAttributes(info);
+
+  // Check model_type 0 (GPT-2)
+  ORT_ENFORCE(parameters_.model_type == 0);
+
+  ONNX_NAMESPACE::GraphProto proto;
+
+  ORT_ENFORCE(info.GetAttr<ONNX_NAMESPACE::GraphProto>("decoder", &proto).IsOK());
+  ORT_IGNORE_RETURN_VALUE(proto);
+}
+
+Status Sampling::SetupSubgraphExecutionInfo(const SessionState& session_state,
+                                            const std::string& attribute_name,
+                                            const SessionState& subgraph_session_state) {
+  const auto& node = Node();
+  if (parameters_.model_type == IGenerationParameters::kModelTypeGpt) {  // GPT-2
+    if (attribute_name == "decoder") {
```

Comment:
>decoder [](http://example.com/codeflow?start=27&length=7)

This PR [https://github.com/microsoft/onnxruntime/pull/13943](https://github.com/microsoft/onnxruntime/pull/13943) handles init_decoder and decoder separately to improve the performance of logits computation of init_decoder. This sampling can leverage the same change.